### PR TITLE
Make `dag_walk::heads_ok()` terminate early in common cases

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/dag_walk.rs [hunk 1]
```diff
@@ -487,16 +487,31 @@ where
...
-    let mut reachable: HashSet<T> = start.iter().cloned().collect();
-    dfs_ok(start.into_iter().map(Ok), id_fn, |node| {
-        let neighbors: Vec<Result<T, E>> = neighbors_fn(node).into_iter().collect();
-        for neighbor in neighbors.iter().filter_map(|x| x.as_ref().ok()) {
-            reachable.remove(neighbor);
+    let mut heads: HashSet<T> = start.iter().cloned().collect();
+    // Do a BFS until we have only one item left in the frontier. That frontier must
+    // have originated from one of the heads, and since there can't be cycles,
+    // it won't be able to eliminate any other heads.
+    let mut frontier: HashSet<T> = heads.clone();
```
Comment (yuja) [inline lib/src/dag_walk.rs:494]: nit: `frontier` doesn't have to be `HashSet`, but I'm not sure if earlier deduplication is better or not.

File: lib/src/dag_walk.rs [hunk 2]
```diff
@@ -487,16 +487,31 @@ where
...
+            })
+            .try_collect()?;
+        for node in &frontier {
+            heads.remove(node);
         }
-        neighbors
-    })
-    .try_for_each(|r| r.map(|_| ()))?;
-    Ok(reachable)
+        frontier.retain(|node| !visited.contains(&id_fn(node)));
```
Comment (yuja) [inline lib/src/dag_walk.rs:512]: nit: maybe `visited` can be updated here?
```rust
frontier.retain(|node| visited.insert(id_fn(node)));
```
In BFS, I think it's common to deduplicate when queueing.

**Review by yuja — APPROVED** [review]
I've added a similar workaround to `topo_order_reverse_lazy`, but it's timestamp-based, and the current `heads_ok()` interface wouldn't be compatible with that. I think BFS is also good.

File: lib/src/dag_walk.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/dag_walk.rs]: Makes sense. Switched to a vector

File: lib/src/dag_walk.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/dag_walk.rs]: Good idea. That avoids calling `id_fn()` twice.

**Pull request merged**